### ISSUE TITLE
Calling `db.get`, `db.update` and `db.merge` in an action on a row that has been deleted throws CBOR error

### ISSUE THREAD

**Issue by rjwebb** [canvasxyz/canvas#385]
Ran into this bug while implementing https://github.com/canvasxyz/canvas/pull/375

It looks like if you delete a row from inside a contract action with `db.delete(...)` then if you then try to call `db.get`, `db.update` or `db.merge` on that row then Canvas throws a CBOR decode error.

It looks like when we write to the `$effects` table from AbstractRuntime, if the `value` is falsy then we write the value and if the value is something else then we `cbor.encode` the value. AFAIK in practice this currently only applies to null values when we have called `db.delete` because in all other cases (create/set/update/merge), `value` is an object.

If we then call `db.get` or `db.update` or `db.merge` on this row in a subsequent action call, `this.getModelValue` (on the AbstractRuntime) is called. It tries to CBOR decode the value (e.g. here and a few other places: https://github.com/canvasxyz/canvas/blob/main/packages/core/src/runtime/AbstractRuntime.ts#L465), but if the value is null, it throws the following error:

```
Error: CBOR decode error: data to decode must be a Uint8Array
      at decodeFirst (file:///Users/bob/code/work/canvas/canvas/node_modules/cborg/lib/decode.js:182:11)
      at Module.decode (file:///Users/bob/code/work/canvas/canvas/node_modules/cborg/lib/decode.js:202:32)
      at Module.decode (file:///Users/bob/code/work/canvas/canvas/node_modules/@ipld/dag-cbor/src/index.js:147:39)
      at FunctionRuntime.getModelValue (file:///Users/bob/code/work/canvas/canvas/packages/core/lib/runtime/AbstractRuntime.js:330:37)
      at async Object.update (file:///Users/bob/code/work/canvas/canvas/packages/core/lib/runtime/FunctionRuntime.js:121:40)
      at async tryUpdatePost (file:///Users/bob/code/work/canvas/canvas/packages/core/test/lib/canvas.test.js:391:21)
      at async FunctionRuntime.execute (file:///Users/bob/code/work/canvas/canvas/packages/core/lib/runtime/FunctionRuntime.js:172:28)
      at async FunctionRuntime.handleAction (file:///Users/bob/code/work/canvas/canvas/packages/core/lib/runtime/AbstractRuntime.js:140:24)
      at async GossipLog.<anonymous> (file:///Users/bob/code/work/canvas/canvas/packages/core/lib/runtime/AbstractRuntime.js:60:24)
      at async GossipLog.apply (file:///Users/bob/code/work/canvas/canvas/packages/gossiplog/lib/AbstractGossipLog.js:234:24)
```

I have a few questions/suggestions about this:

1. We could implement a workaround by checking if the value is `null` when we are trying to decode the row, but why don't we just CBOR encode `null` in this case?
2. If we call `db.get` on a deleted row, do we want to throw an exception? I could see an argument for returning null, e.g. if you write a contract action that checks if a row doesn't exist with a given id.
3. If we call `db.update` or `db.merge` on a deleted row do we also want to throw an exception?

### Reproduction

To reproduce this bug, add the following code to `packages/core/test/canvas.test.ts`, uncomment one of the last three lines (with `tryGetPost`, `tryUpdatePost` and `tryMergePost`) and run the tests.

<details>
<summary>Click to view test code</summary>

```
test("try to access a db row after it has been deleted", async (t) => {
	const wallet = ethers.Wallet.createRandom()
	const app = await Canvas.initialize({
		topic: "com.example.app",
		contract: {
			models: {
				posts: {
					id: "primary",
				},
			},
			actions: {
				async createPost(db, {}: {}, { id }) {
					await db.set("posts", { id })
				},
				async deletePost(db, { id }: { id: string }, {}) {
					await db.delete("posts", id)
				},
				async tryGetPost(db, { id }: { id: string }, {}) {
					await db.get("posts", id)
				},
				async tryMergePost(db, { id }: { id: string }, {}) {
					await db.merge("posts", { id })
				},
				async tryUpdatePost(db, { id }: { id: string }, {}) {
					await db.update("posts", { id })
				},
			},
		},
		signers: [new SIWESigner({ signer: wallet })],
	})
	t.teardown(() => app.stop())

	const { id } = await app.actions.createPost({})
	t.log(`applied action ${id}`)
	await app.actions.deletePost({ id })

	// uncomment one of these to reproduce the error
	// await app.actions.tryGetPost({ id })
	// await app.actions.tryUpdatePost({ id })
	// await app.actions.tryMergePost({ id })
})
```
</details>

**Comment by raykyri** [comment]
Fixing this by returning null if the effect value is null. Test:

```
test("try to access a db row after it has been deleted", async (t) => {
	const wallet = ethers.Wallet.createRandom()
	const app = await Canvas.initialize({
		topic: "com.example.app",
		contract: {
			models: {
				posts: {
					id: "primary",
					foo: "number?",
					bar: "number?",
				},
			},
			actions: {
				async createPost(db, {}: {}, { id }) {
					await db.set("posts", { id, foo: null, bar: null })
				},
				async deletePost(db, { id }: { id: string }, {}) {
					await db.delete("posts", id)
				},
				async tryGetPost(db, { id }: { id: string }, {}) {
					return await db.get("posts", id)
				},
				async tryMergePost(db, { id }: { id: string }, {}) {
					await db.merge("posts", { id, foo: 1, bar: null })
				},
				async tryUpdatePost(db, { id }: { id: string }, {}) {
					await db.update("posts", { id, foo: null, bar: 2 })
				},
			},
		},
		signers: [new SIWESigner({ signer: wallet })],
	})
	t.teardown(() => app.stop())

	// create
	const { id } = await app.actions.createPost({})
	t.deepEqual(await app.db.query("posts", { where: { id } }), [{ id, foo: null, bar: null }])

	// delete
	await app.actions.deletePost({ id })
	t.deepEqual(await app.db.query("posts", { where: { id } }), [])

	// try get
	const { result } = await app.actions.tryGetPost({ id })
	t.deepEqual(result, null)

	// try merge
	await app.actions.tryMergePost({ id })
	t.deepEqual(await app.db.query("posts", { where: { id } }), [{ id, foo: 1, bar: null }])

	// try update
	await app.actions.tryUpdatePost({ id })
	t.deepEqual(await app.db.query("posts", { where: { id } }), [{ id, foo: null, bar: 2 }])
})
```

**Comment by raykyri** [comment]
Closed in 078c93c9. For merge and update, the presence of any ModelValue overwrites `null` so there isn't any special case handling needed there.
